### PR TITLE
DrivAerML: 32k train/eval surface points (vs 50k baseline)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Human team directive: test 32k surface points. Signal from earlier experiment (#3027, mugen): 32k may outperform 50k by reducing noise in sparse point sampling.

## Instructions

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 32000 --drivaerml-eval-surface-points 32000 \
  --max-train-batches 394 \
  --wandb_group dm-surface-points-sweep
```

No `--max-eval-batches`. Report full eval test results.

## Reporting Requirements
- Best val `surface_rel_l2_pct` + epoch
- Test `surface_rel_l2_pct` at best val checkpoint
- W&B run ID

## Baseline

**DrivAerML baselines:**
- val best: `3.997%` @ ep467 (PR #2898, W&B: `bht6h42t`, 4L/512d, AdamW lr=5e-4, T_max=30)
- test best: `6.244%` (PR #2648, 4L/320d, truncated eval — OLDER CONFIG)
- External targets: AB-UPT 3.82% | Transolver-3: 3.71%

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT askeladd:

## Results

### DrivAerML: 32k train/eval surface points (full eval, no --max-eval-batches)

| Metric | 32k (this run) | Baseline (50k, PR #2898) |
|---|---|---|
| **Best val `surface_rel_l2_pct`** | 17.479% @ ep31 | 3.997% @ ep467 |
| **Test `surface_rel_l2_pct`** | 18.501% (at best val ckpt) | 6.244% (older config) |
| **Epochs trained** | 33 | 467 |
| **W&B run** | `8jvflw2w` | `bht6h42t` |
| **External targets** | — | AB-UPT 3.82% / Transolver-3 3.71% |

### Critical throughput finding

The run completed only **33 epochs in 360 minutes** (~10.9 min/epoch) vs the baseline's **467 epochs** (~0.77 min/epoch) — a **14x throughput drop**.

**Root cause:** DrivAerML cases have **~8.14 million surface points** each. Without `--max-eval-batches`:
- **32k eval:** `ceil(8.14M / 32k) = ~255 views/case × 34 val cases = ~8,670 eval batches/epoch`
- **Baseline (50k + `--max-eval-batches 200`):** only **200 eval batches/epoch**

That's **~43x more eval batches**, which consumed nearly the entire training budget. The model was severely epoch-starved — the trough envelope was still descending strongly at cutoff (62.88% → 37.55% → 24.34% → 17.48% over 31 epochs).

### What happened

The 32k hypothesis could not be fairly tested under these conditions. The full eval (no `--max-eval-batches`) meant evaluation dominated wall-clock time, leaving only 33 training epochs vs 467 for the baseline. At matched epoch counts, it's impossible to determine whether 32k outperforms 50k.

The trough-envelope trajectory was healthy and still sharply descending, suggesting the model had significant remaining capacity to improve.

### Command used

```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 32000 --drivaerml-eval-surface-points 32000 \
  --max-train-batches 394 \
  --wandb-group dm-surface-points-sweep \
  --agent askeladd --wandb-name "askeladd/dm-32k-surface-points"
```

### Suggested follow-ups

1. **Re-run 32k with `--max-eval-batches 200`** to match baseline eval budget and get a fair epoch-matched comparison
2. **Alternatively, use `--drivaerml-eval-surface-points 50000 --max-eval-batches 200`** for eval while keeping `--drivaerml-train-surface-points 32000` — this isolates the training signal question from the eval throughput issue
3. Consider eval-every-N-epochs to allow more training within the budget while still getting periodic full-surface evaluation

**Comment by tcapelle** [comment]
ADVISOR: Same epoch-starvation issue — 32k eval points creates ~255 views/case × 34 val cases = 8,670 batches per epoch, consuming nearly all wall-clock time on evaluation. Only 33 epochs completed vs ~467 needed. Please retry with **--max-eval-batches 200** (matching baseline throughput). This is the critical fix to compare 32k vs 50k fairly. Note from the throughput data: once you fix eval overhead, 32k trains faster per epoch — that makes it interesting if it converges to a comparable val metric.

**Comment by tcapelle** [comment]
STUDENT askeladd:

## Results

### Run 2: 32k surface points with --max-eval-batches 200 (throughput fix)

| Metric | 32k (this run) | 50k Baseline (#2898) | Delta |
|--------|---------------|---------------------|-------|
| **Best val `surface_rel_l2_pct`** | **7.558%** (ep128) | **3.997%** (ep467) | +89% worse |
| **Best test `surface_rel_l2_pct`** | **9.217%** (ep128) | N/A | — |
| Epochs completed | 526 (360-min timeout) | 467 | +13% more epochs |
| Time per epoch | ~0.68 min | ~0.77 min | -12% faster |
| External target | — | <3.71% (AB-UPT) | — |

**W&B run:** `hmhzokbo`
**Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 32000 --drivaerml-eval-surface-points 32000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group dm-surface-points-sweep \
  --agent askeladd --wandb-name "askeladd/dm-32k-surface-points-v2"
```

### What happened — NEGATIVE RESULT

**The throughput fix worked** — `--max-eval-batches 200` eliminated the eval bottleneck (0.68 min/epoch vs 0.77 baseline, 12% faster). The model completed 526 epochs in the 360-min budget, exceeding the baseline's 467 epochs.

**However, 32k surface points suffer from catastrophic gradient instability without `--grad-clip`:**

1. **Epochs 1–128:** Normal training. Gradients healthy (norm ~4–10). Val improved from ~15% to best of 7.558% at epoch 128.
2. **Epochs 128–177:** Escalating gradient explosions. Grad norms jumped from ~10 to 174 → 8,237 → 874,163 → millions → billions. Five increasingly severe instability events.
3. **Epochs 177–526:** Terminal instability. Grad norms reached **200+ trillion** at peak. Val oscillated wildly between 25–73%, never recovering. Best-checkpoint weights were already saved at ep128 but the model never improved again.

**Key insight:** The 50k baseline ran the identical architecture (4L/512d, AdamW lr=5e-4, T_max=30, no grad-clip) for 467 epochs without this instability. The 32k surface points produce a fundamentally different gradient landscape that is **unstable without gradient clipping**. This is consistent with the known DrivAerML finding that torch.compile (which changes gradient computation) also diverged at ep454 without `--grad-clip` (PR #2898).

**The best val of 7.558% at ep128 is not a fair comparison to the baseline's 3.997% at ep467** — the 32k model was cut short by instability before reaching comparable epoch depth. The 7.558% represents only ~4 cosine cycles (128/30) vs the baseline's ~15 cycles (467/30).

### Suggested follow-ups

1. **32k + `--grad-clip 1.0` or `--grad-clip 0.5`**: The critical next experiment. Grad clipping would tame the instability and allow fair convergence comparison at matched epoch depth. The 12% throughput advantage makes this worth testing.
2. **32k + grad-clip + EMA**: The new DM EMA baseline uses EMA decay=0.9995 + gc=0.5. Running 32k with both EMA and grad-clip could compound the throughput advantage with EMA smoothing.
3. **40k surface points**: A middle ground — possibly stable without grad-clip while still offering some throughput benefit over 50k.

**Comment by tcapelle** [comment]
ADVISOR: Good progress — eval fix works (0.68 min/epoch, 526 epochs). But 7.558% is confounded by missing EMA+gc. The gradient explosion at ep128 is the exact failure mode that gc=0.5 was designed to prevent.

Please re-run 32k surface points on the full champion platform:

```
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 0.5 --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 32000 --drivaerml-eval-surface-points 32000 --max-train-batches 394 --max-eval-batches 200 --ema-decay 0.9995 --wandb_group dm-surface-points-ema
```

Target: beat 3.833%. The 12% throughput advantage means you'll see more epochs — this is the first fair test of 32k.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3067 because ADVISOR: Closing per issue #3283 directive. 32k surface points is LOWER than 50k baseline — directive says test only denser (64k/96k/128k), not sparser. Reassigning to priority DM lane..
